WIP: Add Exp functionality for decimal

seiflotfy opened the pull request:
Exp computes e^x using the Taylor series to the specified scale and
stores the result in z, which  is also the return value.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5831)

<!-- Reviewable:end -->

seiflotfy committed:
Add Exp functionality for decimal

Exp computes (e^x) (where x = a+b with a being an integer and b < 1)
to the specified scale and stores the result in z, which is also the
return value. We use a combination of exponentiation by squaring for
e^a and taylor series for e^b. e^x = e^a * e^b

seiflotfy commented:
When enabling line  {"41.6572527032084749", "1234567898765432112.2763812"},
It does not converge after 69 iterations. 
```
panic: exp 41.6572527032084749: did not converge after 69 iterations; prev,last result 1234520803724406560.111798045181253501,1234487280474296398.888545559718672482 delta 33523250110161.223252485462581019
```
Am I doing something wrong in the loop?

RaduBerinde commented:
As I already explained in PR #5827, this series converges very slowly for exponents > 1. The number of rounds would need to be a multiple of the exponent, because the term `(x^n / n!)` is still large when x is about the same as n.

You can tune the max # of iterations *per bit of precision* with `newLoop` but this method requires a number of iterations that grows exponentially with the bit size of the exponent.

I suggested a possible solution in the other PR, though again IMO we should look at what other libraries are doing.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-:-KEWZSXHX-b95mbh_9Em:-1464969480)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Actually, it might not be that bad because the bit scale of the output would also grow linearily with the exponent, so the number of rounds might still be linear w.r.t the result scale.  Maybe just setting the # bits ber iteration higher (`newLoop("exp", x, s, 4)`) might work, but I would add tests with larger exponents. Either way it's probably not the best way to implement `Exp`.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-:-KEWaDkCkkEjOZbZ5Z-G:-547328488)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
I agree with @RaduBerinde about checking out what other libraries do to achieve high performance exponentiation. That said, we've taken a lot of inspiration from https://github.com/robpike/ivy while writing this package, and it seems like they also [use a Taylor series](https://github.com/robpike/ivy/blob/e607dc6e37c7eda7053c0d0744102aace83852a4/value/power.go#L63) to calculate exponentials, so that seems like a good sign to me. 

Lets work on playing around with the bits per iteration and making sure this converges on desired precision levels before diving into performance tuning.

---

Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[util/decimal/decimal.go, line 333 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEWkSyMnkejFG9ZV83B:-KEWkSyMnkejFG9ZV83C:1415171919) ([raw file](https://github.com/cockroachdb/cockroach/blob/b03a99415595a82efa0928ce3b9c2b3a6a4a3587/util/decimal/decimal.go#L333)):*
extra space between "which" and "is".

---

*[util/decimal/decimal.go, line 344 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEWmTPLrWgBtNNhe1Ou:-KEWmTPLrWgBtNNhe1Ov:1749640396) ([raw file](https://github.com/cockroachdb/cockroach/blob/b03a99415595a82efa0928ce3b9c2b3a6a4a3587/util/decimal/decimal.go#L344)):*
We can get rid of this and use `decimalOne` instead.

---

*[util/decimal/decimal.go, line 352 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEWkB4hLiJdFin_EKUv:-KEWkB4hLiJdFin_EKUw:-1140069282) ([raw file](https://github.com/cockroachdb/cockroach/blob/b03a99415595a82efa0928ce3b9c2b3a6a4a3587/util/decimal/decimal.go#L352)):*
If this break check is at the end, we can just put it in the for loop itself:

`for loop := ...; !loop.done(z); ; {`

---

*[util/decimal/decimal_test.go, line 379 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEWjTgDiY7_gaoBoJuT:-KEWjTgDiY7_gaoBoJuU:-960302078) ([raw file](https://github.com/cockroachdb/cockroach/blob/b03a99415595a82efa0928ce3b9c2b3a6a4a3587/util/decimal/decimal_test.go#L379)):*
Could you add tests for double scale and a benchmark? Similar to what we've done for each of the other functions.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-:-KEWlDIc0J7gI1U4dQXn:195002490)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
^In that library note the comment "we call it with only small values of x". And see how `floatPower` splits the exponent in the integer and fractional part, using `integerPower` for the integer part. Which is pretty much what I suggested (use the Taylor series only for exponents < 1).

---

Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-:-KEWr0266DPYYwsCQNUE:-1850242305)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Thanks for the update! I think the current solution makes sense mathematically, but we may need to pay extra attention to how much precision we need in the exponents; and the exponentiation in the final loop uses too many steps (we should be doing https://en.wikipedia.org/wiki/Exponentiation_by_squaring#Basic_method). 

Here is a suggestion that I think might be better. The idea is to use the fact that we can trivially compute powers of 10 by using the Scale. Rewrite things as follows:

```
e^x = e^(x ln(10) / ln(10) = 10^(x/ln(10))
```

Compute `x / ln(10)`, then use `Round`(with `RoundDown`) to get the closest integer value (say `I = RoundDown(x / ln(10))`). Get the fractional part as `F = x / ln(10) - I`. 

We can trivially compute the number `10^I` just by setting the scale as `I` (with `NewDec`). 
To compute `10^F`, we turn it back into an exponential `e^(F*ln10)` and use the Taylor Series for `F*ln(10)`.
Finally we multiply these two numbers: `10^F * 10^I = 10^(F+I) = 10^(x/ln(10)) = e^x`

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-:-KEYXIkB4e9o3GT3R31b:-1486144716)*
<!-- Sent from Reviewable.io -->

knz commented on util/decimal/decimal.go:
```diff
@@ -328,3 +328,91 @@ func Log(z *inf.Dec, x *inf.Dec, s inf.Scale) *inf.Dec {
+		return z
+	}
+	xscale := x.Scale()
+	integer := inf.NewDec(u/int64(math.Pow(10, float64(xscale))), 0)
```
You should use integer exponentiation here. 10 and xscale are both ints. No need to go through float.

RaduBerinde commented:
Thanks for working on this! This is starting to look really good!

---

Review status: 0 of 2 files reviewed at latest revision, 15 unresolved discussions.

---

*[util/decimal/decimal.go, line 334 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEal-8XbTyM7SOdJILq:-KEal-8XbTyM7SOdJILr:-2087337165) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal.go#L334)):*
Use better function names. E.g. IntegerPower

---

*[util/decimal/decimal.go, line 339 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEakb0X6fX2KayjTSHk:-KEakb0X6fX2KayjTSHl:-498404934) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal.go#L339)):*
just do neg := (b.Sign() == -1)

---

*[util/decimal/decimal.go, line 346 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEakdDI1YhcjBdXK5_1:-KEakdDI1YhcjBdXK5_2:719835) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal.go#L346)):*
should we verify `ok` here (and maybe panic if !ok?) (I think Nathan knows the best practices around this)

---

*[util/decimal/decimal.go, line 363 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEal9DeGp9EAgyTvQNJ:-KEal9DeGp9EAgyTvQNK:497322799) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal.go#L363)):*
use better function name, mention in the comment that it should only be used for small values of x (preferably between 0 and 1)

---

*[util/decimal/decimal.go, line 403 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEalX-SOtYUIA1qStpK:-KEalX-SOtYUIA1qStpL:-2126921255) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal.go#L403)):*
Add some comments here - explain we are computing `e^(x / 10^xscale)` using `expy` and then raising that to the power of (10^xscale) (or something like that, not sure if I was accurate)

---

*[util/decimal/decimal.go, line 404 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEalxWEowIkS0OCBULe:-KEalxWEowIkS0OCBULf:-1152615572) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal.go#L404)):*
We don't need to use `expx` here, we could just do `Inf.NewDec(1, -xscale)` no?

---

*[util/decimal/decimal_test.go, line 381 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEaoOsRPLfZth9UsJwf:-KEaoOsSXuiaigVyJ3Oa:1815009093) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal_test.go#L381)):*
We need some tests with negative exponents

---

*[util/decimal/decimal_test.go, line 394 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEamOrqVMVJATuQWDZK:-KEamOrqVMVJATuQWDZL:-882155860) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal_test.go#L394)):*
stale comment?

---

*[util/decimal/decimal_test.go, line 395 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEam_WA3E604DbFXDrc:-KEam_WA3E604DbFXDrd:1917681630) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal_test.go#L395)):*
Where did you get the "correct" output of this? I hope not by using the code we are supposed to be testing..

According to wolfram alpha, this is off (starting at the fourth digit after the decimal point)
https://www.wolframalpha.com/input/?i=e%5E41.6572527032084749 (Hit `more digits`)

Also confirmed with http://apfloat.appspot.com/ (expression `exp(n(41.6572527032084749, 1000))`)

---

*[util/decimal/decimal_test.go, line 396 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEamPnMuvzmZDsZ_hPu:-KEamPnMuvzmZDsZ_hPv:1271224426) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal_test.go#L396)):*
Can we add larger tests? something around 300 is still reasonable enough to have a test for. E.g use `exp(n(312.345, 1000))` in http://apfloat.appspot.com/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-:-KEaoCqKsX9k8y0P3IAS:289817521)*
<!-- Sent from Reviewable.io -->

seiflotfy commented:
Review status: 0 of 2 files reviewed at latest revision, 15 unresolved discussions.

---

*[util/decimal/decimal.go, line 333 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEWkSyMnkejFG9ZV83B:-KEeeViet8LUTsVrf-kS:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b03a99415595a82efa0928ce3b9c2b3a6a4a3587/util/decimal/decimal.go#L333)):*
Done.

---

*[util/decimal/decimal.go, line 344 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEWmTPLrWgBtNNhe1Ou:-KEef2PQYR16-311v_WX:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b03a99415595a82efa0928ce3b9c2b3a6a4a3587/util/decimal/decimal.go#L344)):*
Done.

---

*[util/decimal/decimal.go, line 352 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEWkB4hLiJdFin_EKUv:-KEehTQ1wPA2lzhODITt:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b03a99415595a82efa0928ce3b9c2b3a6a4a3587/util/decimal/decimal.go#L352)):*
Done.

---

*[util/decimal/decimal.go, line 406 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KETRk5xtMSFxXvdpaug-r4-406:-KEeiBeJJR_ND0Gu1roV:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ce37addd034ec70e6180a51ab4e3b23d62847a7/util/decimal/decimal.go#L406)):*
Done.

---

*[util/decimal/decimal.go, line 334 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEal-8XbTyM7SOdJILq:-KEeeSq0MSth_sEpPaj7:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal.go#L334)):*
Done.

---

*[util/decimal/decimal.go, line 339 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEakb0X6fX2KayjTSHk:-KEeeTdAvcPwAglYjRAU:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal.go#L339)):*
Done.

---

*[util/decimal/decimal.go, line 403 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEalX-SOtYUIA1qStpK:-KEekIIq5tM5dRyLHak-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal.go#L403)):*
Done.

---

*[util/decimal/decimal.go, line 404 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEalxWEowIkS0OCBULe:-KEeiBBGpZlrrrvDFM8C:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal.go#L404)):*
Done.

---

*[util/decimal/decimal_test.go, line 379 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEWjTgDiY7_gaoBoJuT:-KEekKUsel8AOWii23P2:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b03a99415595a82efa0928ce3b9c2b3a6a4a3587/util/decimal/decimal_test.go#L379)):*
Done.

---

*[util/decimal/decimal_test.go, line 381 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEaoOsRPLfZth9UsJwf:-KEexrw1y3NC_y7ckpxP:733871558) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal_test.go#L381)):*
I am using -7.1, -0.7, -6.6970342501104617 and -0.6931471805599453 is that not enough?

---

*[util/decimal/decimal_test.go, line 394 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEamOrqVMVJATuQWDZK:-KEexceOyKoCgqi9UqJt:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal_test.go#L394)):*
Done.

---

*[util/decimal/decimal_test.go, line 395 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEam_WA3E604DbFXDrc:-KEexV61aI4tMshN-H4q:532431338) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal_test.go#L395)):*
my bad, i was working on fixing the precision which explains why i left the comment, but forgot to revisit it. 
Honest mistake

---

*[util/decimal/decimal_test.go, line 396 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEamPnMuvzmZDsZ_hPu:-KEeyJs8qkOBwArDsIxN:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal_test.go#L396)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

But wait for approval by radu and/or nathan before merging.

---

Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[util/decimal/decimal.go, line 333 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEWkSyMnkejFG9ZV83B:-KEf68UWMywnftrdl18G:-962757097) ([raw file](https://github.com/cockroachdb/cockroach/blob/b03a99415595a82efa0928ce3b9c2b3a6a4a3587/util/decimal/decimal.go#L333)):*
I can still see the double space.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-:-KEf6fBSTXuWO2SaYiiq:-1604998509)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Nice! Looks great, thanks for the contribution!

---

Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[util/decimal/decimal.go, line 403 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEgCcQJ34F-ZF0KcbWR:-KEgCcQJ34F-ZF0KcbWS:1562677793) ([raw file](https://github.com/cockroachdb/cockroach/blob/f46227c1142047c7a32d25608c1acac3d3a04cac/util/decimal/decimal.go#L403)):*
[nit] we no longer use `u` as an integer - we could use Round instead of going through `Unscaled`. Just a suggestion, feel free to ignore.

---

*[util/decimal/decimal.go, line 408 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEgAwBWPY_oQhoRPlLp:-KEgAwBWPY_oQhoRPlLq:-743616783) ([raw file](https://github.com/cockroachdb/cockroach/blob/f46227c1142047c7a32d25608c1acac3d3a04cac/util/decimal/decimal.go#L408)):*
[nit] comments should be wrapped at 80 columns

---

*[util/decimal/decimal.go, line 413 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEgDIckd6M1igiO6gr_:-KEgDIckd6M1igiO6gra:-1008790604) ([raw file](https://github.com/cockroachdb/cockroach/blob/f46227c1142047c7a32d25608c1acac3d3a04cac/util/decimal/decimal.go#L413)):*
[nit] This is hard to understand - we should add a comment explaining the math here, maybe take an example and mention what `n.Scale` is, what `u` is and what `x` is at each step for that example.

---

*[util/decimal/decimal_test.go, line 381 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEaoOsRPLfZth9UsJwf:-KEgCHo9vOQVYAj2XWUX:1849127813) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal_test.go#L381)):*
Ah, my bad, missed them!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-:-KEgCNUzCWzw6GlLh1AX:-398235864)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Looking a lot better! Thanks again @seiflotfy. Just a number of nits now and then we can merge. Also, would you mind adding the benchmark results to the commit message?

---

Review status: 1 of 2 files reviewed at latest revision, 15 unresolved discussions.

---

*[util/decimal/decimal.go, line 344 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEWmTPLrWgBtNNhe1Ou:-KEhJYJ_wqDHvx_8dbV9:-1772672036) ([raw file](https://github.com/cockroachdb/cockroach/blob/b03a99415595a82efa0928ce3b9c2b3a6a4a3587/util/decimal/decimal.go#L344)):*
You can just inline it as well in `n.Add(n, decimalOne)`.

---

*[util/decimal/decimal.go, line 346 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEakdDI1YhcjBdXK5_1:-KEhIT_ajJSAEgByP6bT:642551312) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal.go#L346)):*
Yeah looks like the panic went in. I'm wondering if we should change the signature of this function to 

```
func integerPower(z, a *inf.Dec, b int64, s inf.Scale) *inf.Dec {
```

If we're going to panic on an out of range `b` anyway.

---

*[util/decimal/decimal.go, line 363 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEal9DeGp9EAgyTvQNJ:-KEhKPbsLq-WX36EXJaY:1015954050) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b741c7b7a4b43b57d6b529e472416993e1a7dfb/util/decimal/decimal.go#L363)):*
Also `s/z *inf.Dec, x *inf.Dec/z, x *inf.Dec/`. Same below.

---

*[util/decimal/decimal.go, line 332 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEhHj6iZ-olTE-MWvXB:-KEhHj6iZ-olTE-MWvXC:1090090326) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfcfa710aef175d1041cdd8eba2ce075beeeaef6/util/decimal/decimal.go#L332)):*
These comments should be punctuated. See https://golang.org/doc/effective_go.html#commentary.

---

*[util/decimal/decimal.go, line 341 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEhL7ycm1Pc4DWhAGRo:-KEhL7ycm1Pc4DWhAGRp:-1869686245) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfcfa710aef175d1041cdd8eba2ce075beeeaef6/util/decimal/decimal.go#L341)):*
Are we mutating `b` here? That might be fine since this is a helper function, but double check and please add a comment.

---

*[util/decimal/decimal.go, line 345 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEhHp_P4Mz2lh3KCnH9:-KEhHp_P4Mz2lh3KCnHA:1858718292) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfcfa710aef175d1041cdd8eba2ce075beeeaef6/util/decimal/decimal.go#L345)):*
Why twice?

---

*[util/decimal/decimal.go, line 369 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEhJgL29gZ13zurZxIL:-KEhJgL29gZ13zurZxIM:-1803754107) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfcfa710aef175d1041cdd8eba2ce075beeeaef6/util/decimal/decimal.go#L369)):*
No need to re-allocate `x` because we're never mutating it. The two cases we need to handle are `z == nil` and `z == x`.

---

*[util/decimal/decimal.go, line 378 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEhJwGV9uFIhPT02wq7:-KEhJwGV9uFIhPT02wq8:-1441233296) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfcfa710aef175d1041cdd8eba2ce075beeeaef6/util/decimal/decimal.go#L378)):*
This can be changed to `n := new(inf.Dec)`

---

*[util/decimal/decimal.go, line 390 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEhK67oEgiYxXUq0DS1:-KEhK67oEgiYxXUq0DS2:-1656101872) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfcfa710aef175d1041cdd8eba2ce075beeeaef6/util/decimal/decimal.go#L390)):*
`a` and `b`?

---

*[util/decimal/decimal.go, line 395 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEhKDKJjGvXSjStyRZd:-KEhKDKJjGvXSjStyRZe:1462485865) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfcfa710aef175d1041cdd8eba2ce075beeeaef6/util/decimal/decimal.go#L395)):*
Ditto, we're never modifying `n` so no need to change it. The two cases we need to handle are `z == nil` and `z == n`.

---

*[util/decimal/decimal_test.go, line 379 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEWjTgDiY7_gaoBoJuT:-KEhKhOnCDKd2rb8BUEG:1301031639) ([raw file](https://github.com/cockroachdb/cockroach/blob/b03a99415595a82efa0928ce3b9c2b3a6a4a3587/util/decimal/decimal_test.go#L379)):*
Benchmark looks good! Not seeing a double scale test though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-:-KEhKmvDuM0aoq4wooxW:-1233665927)*
<!-- Sent from Reviewable.io -->

seiflotfy commented:
Review status: 1 of 2 files reviewed at latest revision, 14 unresolved discussions.

---

*[util/decimal/decimal_test.go, line 379 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEWjTgDiY7_gaoBoJuT:-KEh_yoAwk1DBm83T9Fb:150355411) ([raw file](https://github.com/cockroachdb/cockroach/blob/b03a99415595a82efa0928ce3b9c2b3a6a4a3587/util/decimal/decimal_test.go#L379)):*
What do you mean by double scale?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 1 of 2 files reviewed at latest revision, 14 unresolved discussions.

---

*[util/decimal/decimal_test.go, line 379 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEWjTgDiY7_gaoBoJuT:-KEhd6VhMmn9KG4L5wHF:1092090506) ([raw file](https://github.com/cockroachdb/cockroach/blob/b03a99415595a82efa0928ce3b9c2b3a6a4a3587/util/decimal/decimal_test.go#L379)):*
Check out `TestDecimalLogNDoubleScale`, `TestDecimalLog10DoubleScale`, etc.

It should be quick to write, but you'll want to verify that the tests cases all agree with an external verification source (some arbitrary precision calculator online) to a scale of 32.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831)*
<!-- Sent from Reviewable.io -->

seiflotfy committed:
Add double scale test for Exp (TestDecimalExpDoubleScale)

seiflotfy commented:
Done

nvanbenschoten commented:
Review status: 0 of 2 files reviewed at latest revision, 14 unresolved discussions.

---

*[util/decimal/decimal_test.go, line 379 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEWjTgDiY7_gaoBoJuT:-KEhkVUKLFgQRM6gIJ2i:-263162918) ([raw file](https://github.com/cockroachdb/cockroach/blob/b03a99415595a82efa0928ce3b9c2b3a6a4a3587/util/decimal/decimal_test.go#L379)):*
Perfect, the tests look good now. Lets just address those last few comments above and then I think this is good to merge!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831)*
<!-- Sent from Reviewable.io -->

seiflotfy commented:
Review status: 0 of 2 files reviewed at latest revision, 14 unresolved discussions.

---

*[util/decimal/decimal_test.go, line 379 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEWjTgDiY7_gaoBoJuT:-KEhkxdunrmrBRF9BC9o:1537953741) ([raw file](https://github.com/cockroachdb/cockroach/blob/b03a99415595a82efa0928ce3b9c2b3a6a4a3587/util/decimal/decimal_test.go#L379)):*
Which comments? They should all be addressed with my latest commit :D

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice job! @RaduBerinde anything to add, or is this good to merge?

---

Review status: 0 of 2 files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-:-KEhoeAmxfBX-aT08eK-:2060957799)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I have a couple of more comments.

---

Review status: 0 of 2 files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[util/decimal/decimal.go, line 345 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEi2q9zi1o4hUBFymrl:-KEi2qA-WHrNJmQpiDv1:-2096227832) ([raw file](https://github.com/cockroachdb/cockroach/blob/273d342f8a3a22807983862e919524a7c5bcb7c0/util/decimal/decimal.go#L345)):*
Here we are silently ignoring the scale. Which means that if I pass `1.2345` (represented as 12345 with scale 5), I will get back `x^12345`. I think it would make more sense if this function just took `y` as `int64` and let the caller figure out the scaling issues.

---

*[util/decimal/decimal.go, line 392 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEi0FH1XfWC8ED5Yv7B:-KEi0FH1XfWC8ED5Yv7C:2019419945) ([raw file](https://github.com/cockroachdb/cockroach/blob/273d342f8a3a22807983862e919524a7c5bcb7c0/util/decimal/decimal.go#L392)):*
It doesn't look like we modify `nn` anywhere, so we shouldn't need to make a copy of `n`

---

*[util/decimal/decimal.go, line 406 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEhztPrMGZ3qU_s4hGm:-KEhztPrMGZ3qU_s4hGn:-1295012490) ([raw file](https://github.com/cockroachdb/cockroach/blob/273d342f8a3a22807983862e919524a7c5bcb7c0/util/decimal/decimal.go#L406)):*
I am trying to understand this but I'm having some trouble
 - what is `inf.NewDec(1,--nScale())` doing? AFAICT that value isn't actually used anywhere, can't it just be `new(Dec).Round()` ?
 - isn't the Round()` enough to get `integer(n)`? What is the `QuoRound` doing? It looks like we are dividing by `10^nn.Scale()`. Why? I don't think this is needed, and I think things work because it doesn't actually affect anything because we just use `Unscaled` in `integerPower` (see my other comment)

---

*[util/decimal/decimal.go, line 408 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEhyehBWBqoT5diEkL4:-KEhyehCiyLphvBQJx9A:1141130517) ([raw file](https://github.com/cockroachdb/cockroach/blob/273d342f8a3a22807983862e919524a7c5bcb7c0/util/decimal/decimal.go#L408)):*
y = (n -x)    (reads as multiplication otherwise).

---

*[util/decimal/decimal.go, line 412 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEi0S2HUaGBzksDTKia:-KEi0S2HUaGBzksDTKib:-414519201) ([raw file](https://github.com/cockroachdb/cockroach/blob/273d342f8a3a22807983862e919524a7c5bcb7c0/util/decimal/decimal.go#L412)):*
The 1000 here is dubious.. There is no reason for the precision of `ey` to be significantly larger than the one we want (we should be able to use `s`).

---

*[util/decimal/decimal.go, line 413 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEi0xiOzOlWvZhgIEA9:-KEi0xiPjYFWYG_FIq_B:929308740) ([raw file](https://github.com/cockroachdb/cockroach/blob/273d342f8a3a22807983862e919524a7c5bcb7c0/util/decimal/decimal.go#L413)):*
Here it makes sense to have a lot of precision for `e` (since it is the base). But we don't have to compute it every time, we can have `e` be a variable that is initialized once.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-:-KEi3WhlUkulSQCuUIwH:-104185069)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 2 files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[util/decimal/decimal.go, line 406 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEhztPrMGZ3qU_s4hGm:-KEi3y5d1iY4F0R91pnG:-1955345704) ([raw file](https://github.com/cockroachdb/cockroach/blob/273d342f8a3a22807983862e919524a7c5bcb7c0/util/decimal/decimal.go#L406)):*
Sorry, I screwed up the markdown, let me rewrite that last comment:

Isn't the `Round()` enough to get `integer(n)`? What is the `QuoRound` doing? It looks like we are dividing by `10^nn.Scale()`. Why? I don't think this is needed, and I think things work because the scale eventually gets ignored in `integerPower` because we just use `Unscaled` (see my other comment)

 

---

*[util/decimal/decimal.go, line 413 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEi0xiOzOlWvZhgIEA9:-KEi3tCe4MNiMvvghvig:70373968) ([raw file](https://github.com/cockroachdb/cockroach/blob/273d342f8a3a22807983862e919524a7c5bcb7c0/util/decimal/decimal.go#L413)):*
*I meant "be a global variable"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831)*
<!-- Sent from Reviewable.io -->

seiflotfy commented:
*[util/decimal/decimal.go, line 392 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEi0FH1XfWC8ED5Yv7B:-KEi8nJ_2zbslESSlz9c:-11106130) ([raw file](https://github.com/cockroachdb/cockroach/blob/273d342f8a3a22807983862e919524a7c5bcb7c0/util/decimal/decimal.go#L392)):*
somehow if i don't copy it its overwritten.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831)*
<!-- Sent from Reviewable.io -->

seiflotfy commented:
*[util/decimal/decimal.go, line 412 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEi0S2HUaGBzksDTKia:-KEkN_C-AnB8X7pzhMA8:1586110272) ([raw file](https://github.com/cockroachdb/cockroach/blob/273d342f8a3a22807983862e919524a7c5bcb7c0/util/decimal/decimal.go#L412)):*
Because we end up with rounding of rounding.
e.g: 1.9999939 x 1.9999939 != 1.999994 x 1.999994 != 2.0 x 2.0

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831)*
<!-- Sent from Reviewable.io -->

seiflotfy committed:
Clean up decimal.Exp

Make e global variable and simplify the calculation of x and y in
e^(x+y) = e^x * e^y

RaduBerinde commented:
Review status: 0 of 2 files reviewed at latest revision, 19 unresolved discussions.

---

*[util/decimal/decimal.go, line 412 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEi0S2HUaGBzksDTKia:-KEkXU2Le3RXWIBruLe1:130330462) ([raw file](https://github.com/cockroachdb/cockroach/blob/273d342f8a3a22807983862e919524a7c5bcb7c0/util/decimal/decimal.go#L412)):*
Sure but that means we only needs a couple more digits, not 1000. Note that `s` here is higher than the scale that was given to us (it's `+ 2`) and we round down to the original scale below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831)*
<!-- Sent from Reviewable.io -->

seiflotfy commented:
*[util/decimal/decimal.go, line 412 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEi0S2HUaGBzksDTKia:-KEkYPDBO-c5v8DgkaxF:1140466289) ([raw file](https://github.com/cockroachdb/cockroach/blob/273d342f8a3a22807983862e919524a7c5bcb7c0/util/decimal/decimal.go#L412)):*
Then tests will fail.
For e ^ 41.6572527032084749 we 
```
exp: 1234567898765432082.9890763978113354
got: 1234567898765432082.9860183128822650
```

For e ^ 312.345 we
```
exp: 4463853675713824294922499817029570039071067102402155066185430427302882199695129254111120181064791178979160372068542599780002019509758173.2401488061929997
got: 4463853675713824294913818529179243593618232347114359733971740943552074307746566675145829074865196788429958726059479166647483331190648307.9248544996896434
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831)*
<!-- Sent from Reviewable.io -->

seiflotfy commented:
*[util/decimal/decimal.go, line 412 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEi0S2HUaGBzksDTKia:-KEkYPDBO-c5v8DgkaxF:1140466289) ([raw file](https://github.com/cockroachdb/cockroach/blob/273d342f8a3a22807983862e919524a7c5bcb7c0/util/decimal/decimal.go#L412)):*
Then tests will fail.
For e ^ 41.6572527032084749 we 
```
exp: 1234567898765432082.9890763978113354
got: 1234567898765432082.9860183128822650
```

For e ^ 312.345 we
```
exp: 4463853675713824294922499817029570039071067102402155066185430427302882199695129254111120181064791178979160372068542599780002019509758173.2401488061929997
got: 4463853675713824294913818529179243593618232347114359733971740943552074307746566675145829074865196788429958726059479166647483331190648307.9248544996896434
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 2 files reviewed at latest revision, 19 unresolved discussions.

---

*[util/decimal/decimal.go, line 412 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEi0S2HUaGBzksDTKia:-KEkaFeroKvT4XQZGqZI:847847357) ([raw file](https://github.com/cockroachdb/cockroach/blob/273d342f8a3a22807983862e919524a7c5bcb7c0/util/decimal/decimal.go#L412)):*
Ah, you're right. This is a small number and we are multiplying with a potentially huge number `ex` - an error of `10^-s` becomes an error of `10^-s * ex`.

 It still feels hacky to hardcode an arbitrary large value though..

We could add a bit of functionality to `smallPower` so we can pass the factor (`ex`) as well; inside that function we can initialize `tmp` to that value instead of `1`. That way the loop in `smallPower` will automatically end when we reach the precision we want.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831)*
<!-- Sent from Reviewable.io -->

seiflotfy committed:
Simplfy decimal.Exp

Get rid of big scaling hacks and reuse e^x in smallPower

seiflotfy commented:
Review status: 0 of 2 files reviewed at latest revision, 19 unresolved discussions.

---

*[util/decimal/decimal.go, line 412 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEi0S2HUaGBzksDTKia:-KEmIoXdtdkiCnuVrx7R:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/273d342f8a3a22807983862e919524a7c5bcb7c0/util/decimal/decimal.go#L412)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Nice, looking good! LGTM, just a couple of minor nits and it's good to merge.

---

Review status: 0 of 2 files reviewed at latest revision, 18 unresolved discussions.

---

*[util/decimal/decimal.go, line 361 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEmMnI76uwxdiMS9BNx:-KEmMnI9T8eKmRQGt_XK:611710553) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc416d4e6da150735c44c14b71de6053a8924e60/util/decimal/decimal.go#L361)):*
This is not accurately explaining what the function does - it actually computes `z * e^x` no?
Also the name should be maybe `smallExp` since it's not an arbitrary power.

---

*[util/decimal/decimal.go, line 411 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-KEmN5K-g6QPHIFA5RI9:-KEmN5K0Sz4qHOOsi7SQ:1772658675) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc416d4e6da150735c44c14b71de6053a8924e60/util/decimal/decimal.go#L411)):*
stale comment

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-:-KEmN7-0UrWU9wVSS8ay:129423938)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
LGTM

---

Reviewed 1 of 1 files at r9, 1 of 1 files at r13.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831#-:-KEmQxB16I_h6RTUrHQD:676834233)*
<!-- Sent from Reviewable.io -->

seiflotfy commented:
Reviewed 1 of 1 files at r9, 1 of 1 files at r13.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5831)*
<!-- Sent from Reviewable.io -->
